### PR TITLE
Fixing several issues with generics & params and generics

### PR DESCRIPTION
**Pull request by fred-perkins**

### CODE REVIEW

File: src/DSharp.Compiler/Generator/MemberGenerator.cs [hunk 1]
```diff
@@ -295,7 +295,7 @@ private static void GenerateMethod(ScriptGenerator generator, string typeName, M
             if (hasParams)
             {
                 writer.Write($"{DSharpStringResources.ScriptExportMember("paramsGenerator")}(");
-                writer.Write("{0}, ", methodSymbol.Parameters.Count - 1);
+                writer.Write("{0}, ", methodSymbol.Parameters.Count - 1 +(methodSymbol.IsGeneric ? 1 : 0));
```
Comment (isc30) [inline src/DSharp.Compiler/Generator/MemberGenerator.cs]: extract to a meaningful variable name/steps

File: src/DSharp.Compiler/Generator/CodeGenerator.cs [hunk 2]
```diff
@@ -52,6 +52,10 @@ internal static class CodeGenerator
                 foreach (Statement statement in implementation.Statements)
                     StatementGenerator.GenerateStatement(generator, symbol, statement);
             }
+            catch(Exception e)
```
Comment (isc30) [inline src/DSharp.Compiler/Generator/CodeGenerator.cs]: maybe we can make the StatementGenerator an IDisposable and call `EndImplementation()` on Dispose

we would use by doing "using....." and then if an exception gets thrown it properly disposes

File: src/DSharp.Compiler/ScriptModel/Symbols/SymbolSet.cs [hunk 3]
```diff
@@ -900,10 +900,13 @@ public TypeSymbol ResolveType(ParseNode node, ISymbolTable symbolTable, Symbol c
                     typeArguments.Add(argType);
                 }
 
-                TypeSymbol resolvedSymbol = CreateGenericTypeSymbol(templateType, typeArguments);
-                Debug.Assert(resolvedSymbol != null);
+                if(templateType != null)
+                {
+                    TypeSymbol resolvedSymbol = CreateGenericTypeSymbol(templateType, typeArguments);
+                    Debug.Assert(resolvedSymbol != null);
```
Comment (isc30) [inline src/DSharp.Compiler/ScriptModel/Symbols/SymbolSet.cs]: Debug.Assert doesn't give any warning on normal builds

Comment (fred-perkins) [reply src/DSharp.Compiler/ScriptModel/Symbols/SymbolSet.cs]: That's existing code, i'm not that bothered about reworking it right now.

File: src/DSharp.Compiler/Generator/CodeGenerator.cs (hunk 2 above)
Comment (fred-perkins) [reply src/DSharp.Compiler/Generator/CodeGenerator.cs]: Agreed, I probably won't do that work now. That catch is to ensure we actually do show errors to the build as it kept hiding that fact the script output was blowing up.

**Pull request merged**